orttraining/orttraining/training_ops/cuda/nn/batch_norm_grad.cc

```diff
@@ -5,86 +5,131 @@
+      T##_##T1##_##T2,                                                                       \
+      kCudaExecutionProvider,                                                                \
+      (*KernelDefBuilder::Create()).TypeConstraint("T", DataTypeImpl::GetTensorType<T>())    \
+                                   .TypeConstraint("T1", DataTypeImpl::GetTensorType<T1>())  \
+                                   .TypeConstraint("T2", DataTypeImpl::GetTensorType<T2>()), \
+      BatchNormalizationGrad<T, T1, T2>);
+
+template <typename T, typename T1, typename T2>
+Status BatchNormalizationGrad<T, T1, T2>::ComputeInternal(OpKernelContext* ctx) const {
   typedef typename ToCudaType<T>::MappedType CudaT;
+  typedef typename ToCudaType<T1>::MappedType CudaT1;
+  typedef typename ToCudaType<T2>::MappedType CudaT2;
 
   const Tensor* dY = ctx->Input<Tensor>(0);
   const Tensor* X = ctx->Input<Tensor>(1);
   const Tensor* Scale = ctx->Input<Tensor>(2);
   const Tensor* saved_mean = ctx->Input<Tensor>(3);
-  const Tensor* saved_variance = ctx->Input<Tensor>(4);
+  const Tensor* saved_inv_std = ctx->Input<Tensor>(4);
   const TensorShape input_shape = X->Shape();
   const TensorShape channel_shape = saved_mean->Shape();
 
   // no B here, but B has same size as Scale, so can validate inputs for gradient with this substitute
-  ORT_RETURN_IF_ERROR(BatchNormHelper::ValidateInputs(X, Scale, Scale, saved_mean, saved_variance));
+  ORT_RETURN_IF_ERROR(BatchNormHelper::ValidateInputs(X, Scale, Scale, saved_mean, saved_inv_std));
 
   auto dY_data = reinterpret_cast<const CudaT*>(dY->template Data<T>());
   auto X_data = reinterpret_cast<const CudaT*>(X->template Data<T>());
-  auto Scale_data = reinterpret_cast<const CudaT*>(Scale->template Data<T>());
-  auto saved_mean_data = reinterpret_cast<const CudaT*>(saved_mean->template Data<T>());
-  auto saved_variance_data = reinterpret_cast<const CudaT*>(saved_variance->template Data<T>());
+  auto Scale_data = reinterpret_cast<const CudaT1*>(Scale->template Data<T1>());
+  auto saved_mean_data = reinterpret_cast<const CudaT2*>(saved_mean->template Data<T2>());
+  auto saved_inv_std_data = reinterpret_cast<const CudaT2*>(saved_inv_std->template Data<T2>());
 
   auto dX_data = reinterpret_cast<CudaT*>(ctx->Output(0, input_shape)->template MutableData<T>());
-  auto dScale_data = reinterpret_cast<CudaT*>(ctx->Output(1, channel_shape)->template MutableData<T>());
-  auto dBias_data = reinterpret_cast<CudaT*>(ctx->Output(2, channel_shape)->template MutableData<T>());
+  auto dScale_data = reinterpret_cast<CudaT1*>(ctx->Output(1, channel_shape)->template MutableData<T1>());
+  auto dBias_data = reinterpret_cast<CudaT1*>(ctx->Output(2, channel_shape)->template MutableData<T1>());
 
-  const auto alpha = Consts<CudaT>::One;
-  const auto beta = Consts<CudaT>::Zero;
+  const auto alpha = std::is_same<T, MLFloat16>::value ? Consts<float>::One : Consts<CudaT>::One;
```

Comment:
Do we need this condition check here? From the code, for half type, the Zero and One are using float type.
```
template <>
struct Consts<half> {
  static const float Zero;
  static const float One;
};
```

Reply:
Corrected. Thanks Vincent!